### ISSUE TITLE
Should we have an $options parameter

### ISSUE THREAD

**Issue by Nyholm** [php-http/fig-standards#40]
From the discussions [here](https://github.com/php-http/fig-standards/pull/34/files#r132821523). It was suggested to introduce a second parameter like: 

```php
public function sendRequest(RequestInterface $request, int $options = 0): ResponseInterface;
```

Is this something we want? 

Personally, I think the HTTP client should take a request and return a response. ANY response, no matter what the type/code the response has. If it fails to get an response it should throw an exception. In other words: 👎  for an options parameter.

**Comment by WyriHaximus** [comment]
My problem with adding `$options` is that we would have to standardise those options, all of them. For an individual client that isn't really an issue, but for a PSR it is. Would rather suggest using immutability for different options for different requests.

**Comment by clue** [comment]
Before deciding on the interface, is there a consensus on which "options" would be relevant in the first place? This discussion seems to have come up a few times already with no clear output (afaict)?

**Comment by Nyholm** [comment]
I do not think we can define **some** options. It is either all or none. If we do **some** there will always be someone requesting for another option X. 

All options are pretty much `CURLOPT_*`...

**Comment by dbu** [comment]
I am against the options parameter. Its a can of worms and forces all implementations to support all options that we allow. Or we completely ditch interoperability when clients don't have the same behaviour on all of those. And it would promote a very bad practice of splattering the configuration over all code that interacts with the HTTP client.

Rather, people should use middleware/decorators to achieve behaviour like following redirects or throwing exceptions. Some of those behaviours could be outside the specification for a PSR compliant client.

**Comment by joelwurtz** [comment]
Also against, we try it before and we fail hard.

Also after using php-http for more than a year, it's not needed, as 90% of the times you use the same options. (I was before that very skeptical for not having options, long term usage made me realise this was the best decision we ever had)

Different client behavior = different client instance, it works very well.

**Comment by joelwurtz** [comment]
Sorry for double post, but this can also serve as lecture :

 * https://github.com/php-http/httplug/issues/49
 * https://github.com/php-http/httplug/issues/63

**Comment by fbourigault** [comment]
I'm also against using an `$options` parameter. Such parameters are often hiding code smells and poor design.
